### DISCUSSION TITLE
A way to preserve Fern state when reopening/toggling

### DISCUSSION THREAD

**Discussion by iamvoidcoder** [lambdalisue/vim-fern#321] [Ideas]
Hi

It is kinda quiet here in fern.vim discussions section. Not sure this is the right place to ask questions but I'll try anyway.

Is there a way to somehow preserve current Fern browser buffer state when reopening Fern in split window mode or toggling drawer. So that the next time I open/toggle Fern again it is in the same state where I left it last time, including root dir, expanded/collapsed items and cursor position.

Let me explain where I come from. I only just discovered fern.vim and after couple days of toying around I find it to be a super useful plugin. There are just few missing features to make it even better IMHO. For instance, I seems can't figure how to preserve Fern window state after closing then opening it again. 

Say I open a full window Fern via Fern . command. Spend a great deal of time to find the locations and files I need in a huge project. Expand the relevant portions of the tree view, select some initial files then hit open to start editing. Later I need to switch to other files in the same locations. So what are my options? If I do Fern . again I'll be back at square one and have to spend again considerable amount of time to navigate and locate the files. I tried keeping some notes on the paths then use Fern /full/path every time but it takes even longer than simply opening Fern at pwd then navigating to needed locations manually. 

I've been looking through the docs and seems can't find any solution to this problem. The only workaround I can find is to keep drawer all the time open however this too doesn't quite work for me for several reason. First drawer is useless on huge projects with high depth structure. The drawer width is just not enough to have my current working branches open, it doesn't fit unless you make the drawer window real big eating into editing window space at which point I'd rather use a split mode Fern instead.

I suppose I could somehow use a separate "Fern" tab and perhaps make it open files in the first "Editing" tab and remain open but this sounds like an overkill even if possible.

Any ideas please? Did I miss something obvious?

Thanks

**Comment by lambdalisue** [comment]
> It is kinda quiet here in fern.vim discussions section. Not sure this is the right place to ask questions but I'll try anyway.

Welcome! Well, I've recently allowed discussion thus it's quite yet.

> Is there a way to somehow preserve current Fern browser buffer state when reopening Fern in split window mode or toggling drawer. So that the next time I open/toggle Fern again it is in the same state where I left it last time, including root dir, expanded/collapsed items and cursor position.

Yes and no. 

For drawer, yes. Just re-open it by `:Fern . -drawer -toggle` or whatever. Note that `-reveal` would move focus so do not use it.

https://user-images.githubusercontent.com/546312/113479845-3ea25f80-94cc-11eb-97e2-022bd374f29b.mp4

For split, no. Every split buffer will gain different name after opened in design so it's not possible to restore by `:Fern .` command.
Note that you can't restore **but you can get back by `<C-o>` or mark or whatever.**

https://user-images.githubusercontent.com/546312/113479871-5bd72e00-94cc-11eb-9951-31c6ccf64e83.mp4

**Comment by iamvoidcoder** [comment]
Hi. Thanks for the prompt reply!

I tried drawer toggle before asking however it seems to only work if you never leave the pwd. The moment you move up the hierarchy you end up in pwd after showing drawer again. Which most certainly has to do with the fact that you can't invoke Fern without giving it the root path URL. 

I mean, a special case for toggle would be super handy unless this is already doable through some other means, of course. Something like:  

The usual way:
:Fern {url} -toggle ...

Keeping current root:
:Fern -keep_root -toggle ...

Also I'm not quite sure why do you still need to give it the URL when toggling off (hiding)?

Thanks

**Reply by lambdalisue** [reply]
Unfortunately no. We have no idea which is **root** from the command you've suggested.

> Also I'm not quite sure why do you still need to give it the URL when toggling off (hiding)?

Because fern use different buffer for each path in drawer mode and each call in split mode.

**Reply by iamvoidcoder** [reply]
> Unfortunately no. We have no idea which is **root** from the command you've suggested.

Sorry, I must have explained it wrong. I mean, the drawer is already present and showing certain root dir. The drawer certainly 'knows' what the root dir is, does it not? Therefore what is the point of passing it the root if all you want is just hide or show it again without changing whatever root it is displaying atm?

The -keep_root command arg I suggested was to tell the drawer stay pinned to whatever root you are displaying, do not change it.

I suppose the alternative could be a script function saving current drawer root into a variable when toggling off and then passing it back when toggling on. However my Vimscript knowledge is far not enough to put together something that  complex :( 

Something of sorts:

```
let g:fern_last_drawer_root = '.'

function! FernToggleDrawerPreservingRoot() abort

    " Is the drawer visible (about to hide)?
    if (<check_if_fern_drawer_showing>)

        " Backup current drawer root path when closing drawer
        g:fern_last_drawer_root = <fern_drawer_get_current_root>
        
    endif

    " Toggle drawer preserving last displayed root path
    :Fern g:fern_last_drawer_root -drawer -toggle
  
endfunction 
```

**Reply by lambdalisue** [reply]
> The drawer certainly 'knows' what the root dir is, does it not? Therefore what is the point of passing it the root if all you want is just hide or show it again without changing whatever root it is displaying atm?

This is a point you've get wrongly. Drawer is not unique in fern like

1. A drawer for `/`
2. A drawer for `/foo`
3. A drawer for `/foo/bar`

So `-keep-root` does not match with the current design.
